### ISSUE TITLE
Serveless deploy don't adds lambda permissions to alias

### ISSUE THREAD

**Issue by levkaganv** [davidgf/serverless-plugin-canary-deployments#71]
# Bug Report

## Description

After updating to serverless 1.60.4 (from 1.37.1) deployed stack started to return `{"message": "Internal server error"}` for all calls.  

Tried to run "test" from AWS APIGW "Resources"->"Method" view and got: `Thu Dec 26 11:52:32 UTC 2019 : Execution failed due to configuration error: Invalid permissions on Lambda function`

Rolled back to 1.37.1 and deploy worked fine.

After some digging in cloud trails found that serveless 1.60.4 didn't give permissions to the "alias" version of our lambda:

![Screen Shot 2019-12-26 at 1 37 35 PM](https://user-images.githubusercontent.com/41537091/71475092-68f47000-27e7-11ea-98c4-4902444e0b96.png)

## Setup info (parts I thought are relevant):
```yaml
plugins:
  - serverless-plugin-canary-deployments

provider:
  name: aws
  runtime: python3.6

functions:
  - app:
      handler: wsgi_handler.handler
      timeout: 30
      deploymentSettings:
        type: AllAtOnce
        postTrafficHook: tests
        alias: live
```

**Comment by davidgf** [comment]
Hi @levkaganv, thank you very much for the bug report! Is there any chance you could provide me with the CloudFormation templates generated with both v1.60.4 and v1.37.1?

**Comment by levkaganv** [comment]
Hi, I'll try to create a subset of our template that will reproduce the issue.

**Comment by levkaganv** [comment]
Hi @davidgf Attached requested templates:

[templates.zip](https://github.com/davidgf/serverless-plugin-canary-deployments/files/4012671/templates.zip)

**Comment by samcon** [comment]
Hi,

This might be related to this issue:
I'm Running serverless (1.59.2) & serverless-plugin-canary-deployments (0.4.8).
I have a Lambda which is triggered by an ALB. After deploy with canary-deployments, although deployment was fine, I noticed I can't reach my Lambda (ALB returns 502 Bad Gateway).

Looking at CF it seems that canary-deployments points the Original Lambda's permissions to the Candidate function.

So (Also relevant to TestLambdaPermissionRegisterTarget)
```
    "TestLambdaPermissionAlb": {
      "Type": "AWS::Lambda::Permission",
      "Properties": {
        "FunctionName": {
          "Fn::GetAtt": [
            "TestLambdaFunction",
            "Arn"
          ]
        },
        "Action": "lambda:InvokeFunction",
        "Principal": "elasticloadbalancing.amazonaws.com",
        "SourceArn": {
          "Ref": "TestAlbTargetGroupHTTPListener"
        }
      }
    },
```

Becomes
```
    "TempLambdaPermissionAlb": {
      "Type": "AWS::Lambda::Permission",
      "Properties": {
        "FunctionName": {
          "Ref": "TempLambdaFunctionAliasCandidate"
        },
        "Action": "lambda:InvokeFunction",
        "Principal": "elasticloadbalancing.amazonaws.com",
        "SourceArn": {
          "Ref": "TempAlbTargetGroupHTTPListener"
        }
      }
    },
```

I think the Candidate Lambda should probably get the permissions of the Original Lambda *additionally* (in the case that you'll want to access the Candidate from ALB - such as in an E2E test) but I don't think it should it get it instead of the Original Lambda.
I commented out the this.buildPermissionsForAlias and everything works as expected (for my use case).

**Comment by davidgf** [comment]
@levkaganv I've just figured out what the issue is. Basically, [the serverless framework changed the way it builds the Lambda Permissions resource](https://github.com/serverless/serverless/commit/38f6ac125e54d927871b4e5f5b387e0d4c28a6a7#diff-399cbfb87ef7b0f40e9facc24070c61dR16). Prior to `v1.60.2`, the resource would look like this:

```
{
   "HelloLambdaPermissionApiGateway": {
      "Type": "AWS::Lambda::Permission",
      "Properties": {
        "FunctionName": {
          "Fn::GetAtt": [
            "HelloLambdaFunction",
            "Arn"
          ]
        },
   ...
   }
}
```

However, from that version onwards, the Lambda permissions for API Gateway are composed as follows:

```
{
   "HelloLambdaPermissionApiGateway": {
      "Type": "AWS::Lambda::Permission",
      "Properties": {
        "FunctionName": {
          "Fn::Join": [
            ":",
            [
              {
                "Fn::GetAtt": [
                  "HelloLambdaFunction",
                  "Arn"
                ]
              }
            ]
          ]
        },
   ...
   }
}
```

When the plugin amends the permissions to allow API Gateway executing aliased functions, [it looks them up by function name assuming the former structure](https://github.com/davidgf/serverless-plugin-canary-deployments/blob/master/serverless-plugin-canary-deployments.js#L303), and therefore it fails with newer Serverless versions. The fix should be relatively simple, but I cannot give you an ETA (PRs welcome :smile:).

@samcon I'm not sure that your issue is related, I believe that it's just due to ALB events not being supported by the plugin (again, PRs welcome :sweat_smile:)

@dhelms-bw first of all, thank you very much for your PR, it looks good and I'll merge it as soon as I get the time to test it in a real environment. You mention that it fixes the issues discussed here, however I'm failing to see how. Can you shed some light?

**Comment by pda75** [comment]
Thanks.  I did not think to look in the plugin issues!  Will next time.

**Comment by dhelms-bw** [comment]
@davidgf  I apologize. I referenced this issue before I fully understood the source of the problem and I think my PR and this issue are unrelated. The PR I posted fixes a similar issue which has existed before serverless 1.60.2 that exhibits with the same missing alias permission. I'll remove the reference and after seeing your root cause, test my changes against the latest serverless to not bring in any regressions.
Sorry for the confusion :(

**Comment by samcon** [comment]
H,

@davidgf - indeed it seems like a different issue. I'd be happy to create a PR - but I don't think it's an issue with ALB support since it seems to be the intended behaviour in API Gateway. But I'm not sure I understand this behaviour. Why is the permission changed (and not added) to the Alias? Wouldn't the API Gateway lose the permission to invoke the Live Lambda? 

If somehow that makes sense to API Gateway, I don't think it applies to ALB. You probably want:
- ALB -> Live Lambda
- You add the Canary 
- Either Canary is invoked directly by prehook (and thus no ALB permission is needed) or the prehook invokes the Canary through ALB (for instance, via adding separate listener to the Canary Lambda and thus an ALB permission is needed additionally to the Live Lambda)
- If Canary passes validation, LATEST points to Canary, and ALB can invoke it

Thanks!

**Comment by davidgf** [comment]
@samcon this plugin relies on the [Lambda Alias Routing feature](https://docs.aws.amazon.com/lambda/latest/dg/configuration-aliases.html), so traffic gets distributed at Lambda level. That feature allows us diverting the traffic of an alias between two function versions, but `$LATEST` cannot be used. This means that all we need to make sure is that API Gateway methods (or any other event sources that trigger the function you want to deploy gradually) point to the alias instead of `$LATEST` and they have the needed permissions to execute it (if you're curious about how this all works, check out [this blog post](https://hackernoon.com/canary-deployments-in-serverless-applications-b0f47fa9b409)). The plugin is able to manipulate [a number of event sources](https://github.com/davidgf/serverless-plugin-canary-deployments#limitations) so that they target the function alias instead of `$LATEST`, but not all of them are supported. That's the case of ALB based events, the plugin doesn't change them to invoke the alias, and that's probably why you're getting those errors.

**Comment by shaheer-k** [comment]
I have the same issue when I enabled provisionedConcurrency. The API gateway alwasy response with 500 error (Execution failed due to configuration error: Invalid permissions on Lambda function)

Its working if I disable either canary or provisionedcurrency.

When provisionedconcurrency is enabled, then new alias (example: ApiProvConcLambdaAlias) also getting created in addition Live alias. 

Can anyone sugget on how I should  configure the canary deplpysetting when provisionedconcurrency is enabled.

**Comment by davidgf** [comment]
@shaheer-k an existing PR #75 could potentially solve this, but some conflicts need to be resolved first

**Comment by davidgf** [comment]
This issue was solved in `serverless@1.61.0`, so using that version or above should do the trick. However, the plugin is not compatible with functions that have the `provisionedConcurrency` setting enabled, which is something I'll try to fix shortly.
